### ISSUE TITLE
every getindex operation creates a new node

### ISSUE THREAD

**Issue by denizyuret** [denizyuret/Knet.jl#116]
If we repeatedly call an mlp or rnn represented as an array of (boxed) weights w, each time there is a w[i] call  it results in a new node which wastes gpu memory. We should cache getindex results of boxed arrays and let sum_outgrads add the gradients.

**Comment by denizyuret** [comment]
Alternatively we could box each weight w[i] instead of w, thus making getindex(w,i) a non-recorded operation.

**Comment by xukai92** [comment]
Does the new interface (e.g. `Dense` in the `README.md`) also have this issue?

**Comment by denizyuret** [comment]
Yes, this is an issue with AutoGrad: if x is a Param and you use x[i] n
times, getindex(x,i) will get recorded n times. If you instead do xi=x[i]
and use xi n times, getindex(x,i) will be recorded once. This used to be a
bigger issue: every getindex in big hierarchical models used to be
recorded. Now it only effects getindex of Params, e.g. edge cases like
getting columns out of embedding matrices.
